text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1860 @@
+  type would be a type which implements `MetaSized`, and `Pointee`, but not
+  `Sized`. `[usize]` would be referred to as a "`MetaSized` type".
+- "Runtime-sized" types will be used those types whose size is a runtime constant
+  and unknown at compilation time. These would include the scalable vector types
+  mentioned in the motivation below, or those that implement `Sized` but not
+  `const Sized` in the RFC.
+- The bounds on the generic parameters of a function may be referred to simply
+  as the bounds on the function (e.g. "the caller's bounds").
+
+## Acknowledgements
+[acknowledgements]: #acknowledgements
+
+This RFC wouldn't have been possible without the reviews and feedback of
+[@JamieCunliffe][author_jamiecunliffe], [@JacobBramley][ack_jacobbramley],
+[@nikomatsakis][author_nikomatsakis] and [@scottmcm][ack_scottmcm];
+[@eddyb][ack_eddyb] for the `externref` future possibility; the expertise
+of [@compiler-errors][ack_compiler_errors] on the type system and suggesting
+the use of const traits; [@fee1-dead][ack_fee1dead] for reviewing the usage
+of const traits and fixing the ASCII diagrams; and the authors of all of the
+prior art for influencing these ideas.
+
+# Motivation
+[motivation]: #motivation
+
+Introducing a hierarchy of `Sized` and elaborating on the implications of constness
+on `Sized` traits resolves blockers for other RFCs which have had significant interest.
+
+## Runtime-sized types and scalable vector types
+[runtime-sized-types-and-scalable-vector-types]: #runtime-sized-types-and-scalable-vector-types
+
+Rust already supports [SIMD][rfc_simd] (*Single Instruction Multiple Data*),
+which allows operating on multiple values in a single instruction. Processors
+have SIMD registers of a known, fixed length and a variety of intrinsics
+which operate on these registers. For example, x86-64 introduced 128-bit SIMD
+registers with SSE, 256-bit SIMD registers with AVX, and 512-bit SIMD registers
+with AVX-512, and Arm introduced 128-bit SIMD registers with Neon.
+
+As an alternative to releasing SIMD extensions with greater bit widths, Arm and
+RISC-V have vector extensions (SVE/Scalable Vector Extension and the "V" Vector
+Extension/RVV respectively) where the bit width of vector registers depends on the
+CPU implementation, and the instructions which operate these registers are bit
+width-agnostic.
+
+As a consequence, these types are not `Sized` in the Rust sense, as the size of
+a scalable vector cannot be known during compilation, but is a runtime constant.
+For example, the size of these types could be determined by inspecting the value
+in a register - this is not available at compilation time and the value may
+differ between any given CPU implementation. Both SVE and RVV have mechanisms to
+change the system's vector length (up to the maximum supported by the CPU
+implementations) but this is not supported by the proposed ABI for these types.
```

Comment:
What does `size_of::<ScalableVectorType>()` return if SVE/RVV is not available on target cpu? Would it return `0` or just panic?

Reply:
Ideally it wouldn't be possible to do this. Initially that could be by having the types behind `cfg`s, but eventually it'll probably require some other RFCs that allows `target_feature` to be used on structs (but not like in the open `struct_target_features` RFC) so that the type can only be used when the target features are available, or something of that sort.